tests/mock_crust/mod.rs

```diff
@@ -203,48 +214,197 @@ fn multi_split() {
+                }
+            }
+        };
+        nodes_to_add.push(node_to_add);
     }
 
+    //
+    // Arrange
+    // Add new nodes and process until complete
+    //
+    nodes.extend(nodes_to_add.into_iter());
     poll_and_resend(&mut nodes, &mut []);
+
+    //
+    // Assert
+    // Verify that the new nodes are now full nodes part of a section and other invariants.
+    //
+    let non_full_nodes = nodes
+        .iter()
+        .filter(|node| !node.inner.is_node())
+        .map(TestNode::name)
+        .collect_vec();
+    assert!(
+        non_full_nodes.is_empty(),
+        "Should be full node: {:?}",
+        non_full_nodes
+    );
     verify_invariant_for_all_nodes(&mut nodes);
 }
 
+#[test]
+fn simultaneous_joining_nodes_two_sections() {
+    // Create a network with two sections:
+    let network = Network::new(MIN_SECTION_SIZE, None);
+    let nodes = create_connected_nodes_until_split(&network, vec![1, 1], false);
+
+    let prefix_0 = Prefix::default().pushed(false);
+    let prefix_1 = Prefix::default().pushed(true);
+
+    // Relocate in section we were spawned to with a proxy from prefix_0
+    let nodes_to_add_setup = vec![
+        SimultaneousJoiningNode {
+            dst_section_prefix: prefix_0,
+            src_section_prefix: prefix_0,
+            dst_relocation_interval_prefix: None,
+            proxy_prefix: prefix_0,
+        },
+        SimultaneousJoiningNode {
+            dst_section_prefix: prefix_1,
+            src_section_prefix: prefix_1,
+            dst_relocation_interval_prefix: None,
+            proxy_prefix: prefix_0,
+        },
+    ];
+    simultaneous_joining_nodes(network, nodes, &nodes_to_add_setup);
+}
+
+#[test]
+fn simultaneous_joining_nodes_two_sections_switch_section() {
+    // Create a network with two sections:
+    let network = Network::new(MIN_SECTION_SIZE, None);
+    let nodes = create_connected_nodes_until_split(&network, vec![1, 1], false);
+
+    let prefix_0 = Prefix::default().pushed(false);
+    let prefix_1 = Prefix::default().pushed(true);
+
+    // Relocate in section we were not spawned to with a proxy from prefix_0
+    let nodes_to_add_setup = vec![
+        SimultaneousJoiningNode {
+            dst_section_prefix: prefix_0,
+            src_section_prefix: prefix_1,
+            dst_relocation_interval_prefix: None,
+            proxy_prefix: prefix_0,
+        },
+        SimultaneousJoiningNode {
+            dst_section_prefix: prefix_1,
+            src_section_prefix: prefix_0,
+            dst_relocation_interval_prefix: None,
+            proxy_prefix: prefix_0,
+        },
+    ];
+    simultaneous_joining_nodes(network, nodes, &nodes_to_add_setup);
+}
+
+#[test]
+fn simultaneous_joining_nodes_three_section_with_one_ready_to_split() {
+    // TODO: Use same section size once we have a reliable message relay that handle split.
+    // Allow for more route hops otherwise NodeApproaval get losts.
+    let min_section_size = MIN_SECTION_SIZE + 1;
```

Comment:
typo: NodeApproaval -> NodeApproval

could you pls explain bit more regarding why there is lost of NodeApproval, and more route hop will avoid lost? thx

Reply:
I'll try to fix the comment to be a bit clearer.

I'll explain detail here, as I don't think it will be really relevant to the code.
The problem is during split it is possible to loose messages with our current implementation because some node will disconnect. Adding an extra route mean that there are more chances for the message to go through after the split (it is not a full fix but makes it unlikely enough that it will do until we switch to N/3 routing).